File: go/gofind/pkg/gofind/filetypes.go
package gofind

import (
	"database/sql"
	"fmt"
	_ "github.com/mattn/go-sqlite3"
	"strings"
)

type FileType int

const (
	FileTypeNameUnknown = "unknown"
	FileTypeNameArchive = "archive"
	FileTypeNameAudio   = "audio"
	FileTypeNameBinary  = "binary"
	FileTypeNameCode    = "code"
	FileTypeNameFont    = "font"
	FileTypeNameImage   = "image"
	FileTypeNameText    = "text"
	FileTypeNameVideo   = "video"
	FileTypeNameXml     = "xml"
)

const (
	FileTypeUnknown FileType = iota
	FileTypeArchive FileType = iota
	FileTypeAudio   FileType = iota
	FileTypeBinary  FileType = iota
	FileTypeCode    FileType = iota
	FileTypeFont    FileType = iota
	FileTypeImage   FileType = iota
	FileTypeText    FileType = iota
	FileTypeVideo   FileType = iota
	FileTypeXml     FileType = iota
)

type FileTypes struct {
	db                  *sql.DB
	extTypeCache        map[string]FileType
	nameTypeCache       map[string]FileType
	nameTypeCacheLoaded bool
}

// used for unmarshalling
type JsonFileType struct {
	Type       string
	Extensions []string
	Names      []string
}

type JsonFileTypes struct {
	FileTypes []*JsonFileType
}

func NewFileTypes() *FileTypes {
	config := NewFindConfig()
	var fileTypes FileTypes
	db, err := sql.Open("sqlite3", fmt.Sprintf("file:%v?mode=ro", config.XFINDDB))

	if err != nil {
		//log.Fatal(err)
		return &fileTypes
	}
	fileTypes.db = db
	fileTypes.extTypeCache = make(map[string]FileType)
	return &fileTypes
}

func (ft *FileTypes) GetFileTypesForQueryAndParams(query string, params []any) map[string]FileType {
	results := make(map[string]FileType)
	rows, err := ft.db.Query(query, params...)
	if err != nil {
		return results
	}
	defer rows.Close()
	for rows.Next() {
		var key string
		var fileTypeId int
		if err := rows.Scan(&key, &fileTypeId); err != nil {
			return results
		}
		results[key] = FileType(fileTypeId - 1)
	}
	return results
}

func (ft *FileTypes) loadNameTypeCache() {
	ft.nameTypeCache = ft.GetFileTypesForQueryAndParams("SELECT name, file_type_id FROM file_name", []any{})
	ft.nameTypeCacheLoaded = true
}

func (ft *FileTypes) GetFileTypeForQueryAndParams(query string, params []any) FileType {
	stmt, err := ft.db.Prepare(query)
	if err != nil {
		return FileTypeUnknown
	}
	defer stmt.Close()
	var fileTypeId int
	err = stmt.QueryRow(params...).Scan(&fileTypeId)
	if err != nil {
		return FileTypeUnknown
	}
	return FileType(fileTypeId - 1)
}

func (ft *FileTypes) GetFileTypeForFileName(fileName string) FileType {
	if !ft.nameTypeCacheLoaded {
		ft.loadNameTypeCache()
	}
	if fileType, found := ft.nameTypeCache[fileName]; found {
		return fileType
	}
	//query := "select file_type_id from file_name where name = ?"
	//return ft.GetFileTypeForQueryAndParams(query, []any{fileName})
	return FileTypeUnknown
}

func (ft *FileTypes) GetFileTypeForExtension(ext string) FileType {
	if ext == "" {
		return FileTypeUnknown
	}
	if fileType, found := ft.extTypeCache[ext]; found {
		return fileType
	}
	query := "select file_type_id from file_extension where extension = ?"
	fileType := ft.GetFileTypeForQueryAndParams(query, []any{ext})
	ft.extTypeCache[ext] = fileType
	return fileType
}

func (ft *FileTypes) GetFileType(file string) FileType {
	fileTypeForFileName := ft.GetFileTypeForFileName(file)
	if fileTypeForFileName != FileTypeUnknown {
		return fileTypeForFileName
	}
	return ft.GetFileTypeForExtension(GetExtension(file))
}

func GetFileTypeForName(name string) FileType {
	lname := strings.ToLower(name)
	switch lname {
	case FileTypeNameArchive:
		return FileTypeArchive
	case FileTypeNameAudio:
		return FileTypeAudio
	case FileTypeNameBinary:
		return FileTypeBinary
	case FileTypeNameCode:
		return FileTypeCode
	case FileTypeNameFont:
		return FileTypeFont
	case FileTypeNameImage:
		return FileTypeImage
	case FileTypeNameText:
		return FileTypeText
	case FileTypeNameVideo:
		return FileTypeVideo
	case FileTypeNameXml:
		return FileTypeXml
	default:
		return FileTypeUnknown
	}
}

func GetNameForFileType(fileType FileType) string {
	switch fileType {
	case FileTypeArchive:
		return FileTypeNameArchive
	case FileTypeAudio:
		return FileTypeNameAudio
	case FileTypeBinary:
		return FileTypeNameBinary
	case FileTypeCode:
		return FileTypeNameCode
	case FileTypeFont:
		return FileTypeNameFont
	case FileTypeImage:
		return FileTypeNameImage
	case FileTypeText:
		return FileTypeNameText
	case FileTypeVideo:
		return FileTypeNameVideo
	case FileTypeXml:
		return FileTypeNameXml
	default:
		return FileTypeNameUnknown
	}
}

func (ft *FileTypes) IsArchiveFile(file string) bool {
	return ft.GetFileType(file) == FileTypeArchive
}

func (ft *FileTypes) IsAudioFile(file string) bool {
	return ft.GetFileType(file) == FileTypeAudio
}

func (ft *FileTypes) IsBinaryFile(file string) bool {
	return ft.GetFileType(file) == FileTypeBinary
}

func (ft *FileTypes) IsCodeFile(file string) bool {
	return ft.GetFileType(file) == FileTypeCode
}

func (ft *FileTypes) IsFontFile(file string) bool {
	return ft.GetFileType(file) == FileTypeFont
}

func (ft *FileTypes) IsImageFile(file string) bool {
	return ft.GetFileType(file) == FileTypeImage
}

func (ft *FileTypes) IsTextFile(file string) bool {
	fileType := ft.GetFileType(file)
	return fileType == FileTypeText || fileType == FileTypeCode || fileType == FileTypeXml
}

func (ft *FileTypes) IsVideoFile(file string) bool {
	return ft.GetFileType(file) == FileTypeVideo
}

func (ft *FileTypes) IsXmlFile(file string) bool {
	return ft.GetFileType(file) == FileTypeXml
}

func (ft *FileTypes) IsUnknownFile(file string) bool {
	return ft.GetFileType(file) == FileTypeUnknown
}

func ContainsFileType(fileTypes []FileType, fileType FileType) bool {
	for _, ft := range fileTypes {
		if fileType == ft {
			return true
		}
	}
	return false
}

File: elixir/exfind/lib/filetypes.ex
defmodule ExFind.FileType do
  @moduledoc """
  Documentation for `ExFind.FileType`.
  """

  defstruct type_name: "", extensions: [], names: []

  def new(args), do: __struct__(args)
end

defmodule ExFind.FileTypes do
  @moduledoc """
  Documentation for `ExFind.FileTypes`.
  """

  alias ExFind.FileUtil

  @file_types [:unknown, :archive, :audio, :binary, :code, :font, :image, :text, :video, :xml]

  defstruct [:conn, :file_types]

  def new() do
    {:ok, conn} = Exqlite.Sqlite3.open(ExFind.Config.xfind_db_path, [:readonly])
    :ets.new(:ext_type_cache, [:set, :public, :named_table])
    :ets.new(:name_type_cache, [:set, :public, :named_table])
    self = __struct__([conn: conn, file_types: @file_types])
    load_name_type_cache(self)
    self
  end

  def get_file_type_for_name(name) do
    file_types = @file_types
    name_atom = if is_atom(name), do: name, else: String.downcase(name) |> String.to_atom()
    if Enum.member?(file_types, name_atom), do: name_atom, else: :unknown
  end

  def get_name_type_map_for_statement(file_types, statement, name_type_map) do
    case Exqlite.Sqlite3.step(file_types.conn, statement) do
      {:row, [name, file_type_id]} ->
        file_type = Enum.at(file_types.file_types, file_type_id - 1)
        get_name_type_map_for_statement(file_types, statement, Map.put(name_type_map, name, file_type))
      :done -> name_type_map
    end
  end

  def get_file_types_for_query_and_params(file_types, query, params) do
    {:ok, statement} = Exqlite.Sqlite3.prepare(file_types.conn, query)
    :ok = Exqlite.Sqlite3.bind(file_types.conn, statement, params)
    name_type_map = get_name_type_map_for_statement(file_types, statement, %{})
    :ok = Exqlite.Sqlite3.release(file_types.conn, statement)
    name_type_map
  end

  def load_name_type_cache(file_types) do
    query = "SELECT name, file_type_id FROM file_name;"
    name_type_map = get_file_types_for_query_and_params(file_types, query, [])
    Enum.each(name_type_map, fn {name, file_type} -> :ets.insert(:name_type_cache, {name, file_type}) end)
  end

  def get_file_type_for_query_and_params(file_types, query, params) do
    {:ok, statement} = Exqlite.Sqlite3.prepare(file_types.conn, query)
    :ok = Exqlite.Sqlite3.bind(file_types.conn, statement, params)
    file_type = case Exqlite.Sqlite3.step(file_types.conn, statement) do
      {:row, [file_type_id]} -> Enum.at(file_types.file_types, file_type_id - 1)
      :done -> :unknown
    end
    :ok = Exqlite.Sqlite3.release(file_types.conn, statement)
    file_type
  end

  def get_db_file_type_for_file_name(file_types, file_name) do
    case :ets.lookup(:name_type_cache, file_name) do
      [{_, file_type}] ->
        # IO.puts("Found file type in cache for #{file_name}: #{file_type}")
        file_type
      [] ->
        # query = "SELECT file_type_id FROM file_name WHERE name = ?1;"
        # file_type = get_file_type_for_query_and_params(file_types, query, [file_name])
        # file_type
        :unknown
    end
  end

  def get_db_file_type_for_extension(file_types, file_ext) do
    if file_ext == "" do
      :unknown
    else
      case :ets.lookup(:ext_type_cache, file_ext) do
        [{_, file_type}] ->
          file_type
        [] ->
          query = "SELECT file_type_id FROM file_extension WHERE extension = ?1;"
          file_type = get_file_type_for_query_and_params(file_types, query, [file_ext])
          :ets.insert(:ext_type_cache, {file_ext, file_type})
          file_type
      end
    end
  end

  def get_file_type_for_file_name(file_types, file_name) do
    case get_db_file_type_for_file_name(file_types, file_name) do
      :unknown -> get_db_file_type_for_extension(file_types, FileUtil.get_extension(file_name))
      file_type -> file_type
    end
  end

  def get_file_type_for_path(file_types, file_path) do
    get_file_type_for_file_name(file_types, Path.basename(file_path))
  end

  def archive_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :archive
  end

  def audio_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :audio
  end

  def binary_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :binary
  end

  def code_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :code
  end

  def font_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :font
  end

  def image_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :image
  end

  def text_file_name?(file_types, file_name) do
    file_type = get_file_type_for_file_name(file_types, file_name)
    Enum.any?([:text, :code, :xml], fn t -> t == file_type end)
  end

  def video_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :video
  end

  def xml_file_name?(file_types, file_name) do
    get_file_type_for_file_name(file_types, file_name) == :xml
  end
end
